scalyr_agent/configuration.py

```diff
@@ -386,60 +386,117 @@ def __verify_api_keys(self):
         """
         Verify all api key config entries from the "api_keys" list in config.
         """
-        api_keys = self.__config.get_json_array("api_keys")
-        default_key_config = None
-        for key_config in api_keys:
-            api_key = key_config.get("api_key", none_if_missing=True)
-            # if the 'api_key' from this entry equals to the 'main' api key,
-            # than this api key entry is considered as default.
-            if api_key is None or api_key == self.api_key:
-                if default_key_config is None:
-                    default_key_config = key_config
-                else:
-                    # if there are more than one api key entries that also can be considered as a default,
-                    # then raise a duplication error
-                    raise BadConfiguration(
-                        "The default api key is redefined. Please use only unique api keys."
-                        "The entries without 'api_key' field are also considered as default",
-                        "api_keys",
-                        "apiKeyDuplication",
-                    )
 
-        unique_keys = set()
-        other_api_key_configs = list()
-        # If there are multiple api key entries with the same api_key, then the duplication error is raised too.
-        for key_config in api_keys:
-            api_key = key_config.get("api_key", none_if_missing=True)
-            if api_key in unique_keys:
-                key_id = key_config["id"]
+        api_keys = list(self.__config.get_json_array("api_keys"))
+
+        unique_api_ids = {}
+        # Apply other defaults to all api key entries
+        for i, api_key_entry in enumerate(api_keys):
+            self.__verify_api_keys_entry_and_set_defaults(api_key_entry)
+            api_key_id = api_key_entry["id"]
+            if api_key_id in unique_api_ids:
                 raise BadConfiguration(
-                    "The api key with api_key_id '%s' is redefined in next entries. Please use only unique api keys."
-                    % key_id,
+                    "The entry field 'api_key_id' of the list 'api_keys' has to be unique.",
```

Comment:
We should include the specific value of `api_key_id` that caused a duplicate in the error message.

Also, because you set `"default"` as the id for any entry that doesn't have an api key set, this error message could be confusing if there are multiple entries that do not have the `id` field specified.   Maybe you should special case the error message for `"default"` and say either there are multiple entries that have `"default"` or mulitple entries that do not have `id` defined.

Reply:
Sorry, I forgot to answer this comment https://github.com/scalyr/scalyr-agent-2/pull/664#discussion_r537845613
but I've made exactly as said there. 

> Also, because you set "default" as the id for any entry that doesn't have an api key set, ...
Now it just throws an error if the `api_key` file is missing. The `api_key` and `id` fields are required now

It is also impossible to have a duplication error with two `default` ids unless the user does not add them explicitly, because nothing is created automatically in this case and nothing can confuse the user.
If it adds only one default key, then everything is fine. In this case, we just do not create the default key automatically.

Does that make sense to you?

PS: I've also changed the error message to
 
_"The api key #0 already has id 'some id' but it has also been re-used by the api key #1. Api key id's must remain unique."_

`#0`, `#1` - positions in the `api_keys` list

Reply:
I don't think using the indexes will be that meaningful in all cases.  Remember, those key entries could be created by joining together the information in across multiple `.json` files.  So, I think we should just say there are entries with duplicate ids and give the id name.

Ok, I saw that there is code down below in the `verify` method that adds in `default` if the `id` field is not specified.  Maybe that was dead code and you meant to remove it.  But yes, that makes sense.